### ISSUE TITLE
Automatic fixing by plugins [REPLACEMENT ISSUE]

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#545]
The original issue

	Id: 286
	Title: Automatic fixing by plugins

could not be created.
This is a dummy issue, replacing the original one. It contains everything but the original issue description. In case the gitlab repository is still existing, visit the following link to show the original issue:

TODO

**Comment by asottile** [comment]
In GitLab by @ambv on Dec 21, 2016, 02:23

I recommended against mutating the FST as response to raised errors because it causes surprising effects. One fixer can trigger another warning. If you executed the fixers recursively in this case, we cannot guarantee it will ever finish (fixer cycles, etc.).

[yapf](https://github.com/google/yapf) took a more holistic approach: unwrap lines, permutate wrapping while applying penalty points for things the user doesn't like, choose the formatting with the least penalty points. This is the algorithm used by clang-format. This gives much better results but still causes surprises every now and then. Specifically, applying correct weights in config becomes an art form.

gofmt is more simplistic, AFAICT it has limited configuration, similar to JSON pretty printers, etc.

Don't get me wrong, auto-fixers are an awesome productivity boost, I'd very much like to see them. It's just that it's a hard problem and in practice the result is semi-automatic at best with FST, or semi-elegant with applied weights or pretty printers.

**Comment by asottile** [comment]
In GitLab by @joe-gordon0 on Jan 6, 2017, 19:06

yapf uses lib2to3 underneath, which has a big warning in the docs: https://docs.python.org/2/library/2to3.html#module-lib2to3. I am not sure if someone has forked lib2to3 or if the warning is outdated.

having played around with [autopep8](https://github.com/hhatto/autopep8) I have seen some of the issues @ambv describes. And I think semi-automatic is what we should aim for.

@sigmavirus24 how many pep8 rules have low risk of causing other issues and have few ways of fixing? Perhaps aiming to solve those cases first may make things a little easier?

Or we could just embrace yapf and be done with it?

**Comment by asottile** [comment]
In GitLab by @ambv on Jan 7, 2017, 11:16

lib2to3 is very unlikely to change anymore. Also, the usage of it by yapf is limited mostly to unwinding lines.

There is no inherently "safe" subset of rules, whatever changes line lengths or affects other formatting might trigger warnings that weren't there before. Not to mention people running custom plugins that may have additional checks. So, the way to introduce "safe" rules would be to suggest auto-fixes to the user only after we applied them temporarily and confirmed no *additional* warnings are generated. The complication here is that we'd have to track line and column changes to differentiate between *existing* and *additional* warnings. Better yet, those *additional* warnings might be triggered outside of the lines that are fixed, for example when cyclomatic complexity is increased or a name used before isn't anymore.

If it sounds like I tried this in the past, it's because I did ;-) In the end it seems yapf is simply much more predictable. Funnily, it also sometimes generates code that disagrees with pycodestyle if the configuration isn't in complete sync :D

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 23, 2017, 04:38

> Or we could just embrace yapf and be done with it?

It sounds like yapf has improved. What I think is missing, however, is the ability to have one configuration for the autofixer and flake8. Most people already have a project style-guide in the way of select and ignore rules for Flake8.

Perhaps the best thing for us to do would be to have a way to generate a yapf config from Flake8's?

**Comment by asottile** [comment]
In GitLab by @joe-gordon0 on Mar 18, 2017, 12:08

@sigmavirus24 just using yapf sounds like a good idea, less re-inventing the wheel needed.

**Comment by asottile** [comment]
In GitLab by @taion on Nov 23, 2017, 24:27

Autofixing in a linter and autoformatting code aren't the same thing, and I think keeping in mind the differences between the two simplifies the problem a bit.

In a nutshell, autofixing in a linter should _patch_ code, but it doesn't need to be comprehensive. By contrast, autoformatting tools generally _rewrite_ code entirely, and correspondingly must be comprehensive.

To take an example of the former, consider ESLint in the JavaScript ecosystem. ESLint is the most popular linting tool in that ecosystem, and substantially all modern projects now use ESLint for linting.

In ESLint, only a subset of the rules are fixable. These are exactly those rules that can be fixed unambiguously. Running ESLint with `--fix` will greedily fix things as needed, but it makes no guarantees that the fixes don't impose other rule violations.

To take a concrete example, ESLint has a fixable [`array-bracket-spacing`](https://eslint.org/docs/rules/array-bracket-spacing) rule that controls how array brackets are spaced. It has a non-fixable [`max-len`](https://eslint.org/docs/rules/max-len) rule that controls how long lines should be.

If I configure `array-bracket-spacing` to mandate spaces in array brackets, then it's possible that running ESLint with `--fix` will insert spaces such that my line now violates the `max-len` rule.

In principle, per the above, that sounds like a problem – but actually it isn't. In practice, the DX there is fine. Using `eslint --fix` in this manner will have (in the general case) saved me time, cleaned up my code style, and only in a few cases will force me to make another code style fix – but I'll have had to make that fix anyway.

This is less comprehensive, but in practice that's fine. If I wanted to have all of my code be automatically formatted, I'd use an autoformatting tool and not worry about linting for code style. In the case where I've chosen to lint for code style, greedy local patches are just fine for autofixing.

Something like using YAPF to rewrite all of my code is unnecessary and probably not desirable – if I thought YAPF were beneficial, I'd already be using it.

**Comment by asottile** [comment]
In GitLab by @taion on Nov 23, 2017, 24:34

And note that the API for rules applying fixes in ESLint mirrors what I said above – it's geared around rules applying small changes to the ~~FST~~AST: https://github.com/eslint/eslint/blob/f39ffe713dc355b6a2db1ee73a01a7267bea18e8/lib/rules/array-bracket-spacing.js#L126-L128.

It's not exactly analogous to what would make sense here, since ESLint has a more structured API with independent rules rather than the more arbitrary plugin scheme here where the only common ground is in reported error codes, but it's a neat, practical example.

I believe RuboCop works much the same way.

**Comment by asottile** [comment]
In GitLab by @taion on Nov 23, 2017, 24:39

Oops, misspoke above on "FST" in the ESLint fixer example. That example is actually working with the AST. Given the local scope of the fix changes there, the AST is sufficient.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Nov 23, 2017, 05:04

> In ESLint, only a subset of the rules are fixable.

Right. In my opinion, there are a subset of violations that Flake8 could fix on its own (e.g., unused imports). It was never my intention to have Flake8 fix all possible style and lint issues. Some people would like it if that were possible, but alas its not a goal. I do think having this be pluggable would be preferable so people can build their own ways of fixing different rules.

That aside, Python's AST is not something we can round-trip. We can't go from `source -> AST -> source` while preserving indentation, comments, line-breaks, etc. Until we can do that this is pretty much a moot conversation. If there's an alternative that allows for that, is at least as fast as the standard library `ast` module, and has the same API (so plugins don't notice a difference) as the standard library then I'd happily consider replacing it to get closer to this functionality.

**Comment by asottile** [comment]
In GitLab by @taion on Nov 23, 2017, 08:47

You don't need to fully round-trip through the AST. It's sufficient to record a set of edits in text space based on locations in the AST (or from looking at the physical code). You need to do a little bit of bookkeeping in case you have multiple overlapping edits, but otherwise it's not too bad.

This looks like it's essentially how autopep8 works. For things against the AST, things like deletions could just be represented as "delete everything until the next parsed token".

The implementation in ESLint looks like this:
- https://github.com/eslint/eslint/blob/v4.11.0/lib/util/rule-fixer.js (records fixes)
- https://github.com/eslint/eslint/blob/v4.11.0/lib/util/source-code-fixer.js (applies fixes)

**Comment by asottile** [comment]
In GitLab by @taion on Nov 23, 2017, 10:17

Also, while this isn't relevant here given my point above, it's not the case that you need to operate on a CST (or FST) to round-trip syntax changes to source code.

Broadly, instead you can just track modifications to the AST, and then only modify e.g. lines of code that were affected by e.g. the automatic refactoring. If you can keep track of which parts of the AST did not see changes, you can just not modify the source code for those sections of the tree.

In the JavaScript ecosystem, this is the approach taken by [Recast](https://github.com/benjamn/recast), which underlies Facebook's [codemod](https://github.com/facebook/codemod), which FB have claimed that they use extensively internally to do large-scale refactorings, without mangling their code formatting too badly. (Although they've since switched to a heuristics-based autoformatter that in practice gives very good results for most JavaScript code.)

I'm not personally familiar with RedBaron, but I've used codemod to provide (and consume) refactorings for libraries I've maintained in the past (https://github.com/reactjs/rackt-codemod), and it was quite easy to work with, and in practice codemod via Recast did a very good job of not touching my code formatting except for the bits that it needed to change the code.

So to reiterate, the example that most immediately jumps to my mind for a battle-tested, production-grade code refactoring tool that is routinely used against a very large codebase actually works in AST space, and doesn't expose anything CST-related to people writing code transforms.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 20, 2018, 16:25

Closing this off with the existence of black which works really well and is being used on this project.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 20, 2018, 16:25

closed

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 2, 2020, 16:37

changed the description

**Comment by asottile** [comment]
flake8 doesn't have any of the facilities to enable source rewriting and will not be growing the complexity to do so

it's not a simple task to do so, requiring some sort of standardized CST (of which there is no functioning one which keeps up with the syntax changes in cpython) or alternative approach (tokenize-rt+ast for example -- though it's much more work to implement fixers there)

black solves most of the built-in issues of pycodestyle which was the original intent of this issue (to subsume something like autopep8 / black into flake8)

**Comment by sigmavirus24** [comment]
To be clear, when I started this I was thinking that Flake8 would provide default pep8/pyflakes fixes that users could override/replace with plugins like the rest of the Flake8 ecosystem. I also imagined plugins could define their own fixers. There was never very much interest though, so we ripped out the bits that I had plumbed in a bit to start to provide a semblance of structure.

Like @asottile said though, this is incredibly thorny and not something we want to invest time in designing, building, or maintaining.
